src/target/riscv/riscv-013.c

```
@@ -1556,24 +1570,77 @@ static int wait_for_authbusy(struct target *target, uint32_t *dmstatus)
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKS, riscv_ebreaks && riscv_supports_extension(target, 'S'));
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKU, riscv_ebreaku && riscv_supports_extension(target, 'U'));
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKVS, riscv_ebreaku && riscv_supports_extension(target, 'V'));
+	dcsr = set_field(dcsr, CSR_DCSR_EBREAKVU, riscv_ebreaku && riscv_supports_extension(target, 'V'));
+	if (dcsr != original_dcsr &&
+			riscv_set_register(target, GDB_REGNO_DCSR, dcsr) != ERROR_OK)
+		return ERROR_FAIL;
+	info->dcsr_ebreak_set = true;
+	return ERROR_OK;
+}
+
+static int halt_set_dcsr_ebreak(struct target *target)
+{
+	RISCV_INFO(r);
+	RISCV013_INFO(info);
+	LOG_TARGET_DEBUG(target, "Halt to set DCSR.ebreak*");
+
+	/* Remove this hart from the halt group.  This won't work on all targets
+	 * because the debug spec allows halt groups to be hard-coded, but I
+	 * haven't actually encountered those in the wild yet.
+	 *
+	 * There is a possible race condition when another hart halts, and
+	 * this one is expected to also halt because it's supposed to be in the
+	 * same halt group. Or when this hart is halted when that happens.
+	 *
+	 * A better solution might be to leave the halt groups alone, and track
+	 * why we're halting when a halt occurs. When there are halt groups,
+	 * that leads to extra halting if not all harts need to set dcsr.ebreak
+	 * at the same time.  It also makes for more complicated code.
+	 *
+	 * The perfect solution would be Quick Access, but I'm not aware of any
+	 * hardware that implements it.
+	 *
+	 * We don't need a perfect solution, because we only get here when a
+	 * hart spontaneously resets, or when it powers down and back up again.
+	 * Those are both relatively rare. (At least I hope so. Maybe some
+	 * design just powers each hart down for 90ms out of every 100ms)
+	 */
+
+	if (info->haltgroup_supported && set_group(target, NULL, 0, HALT_GROUP) != ERROR_OK)
+		return ERROR_FAIL;
```

There is no error print inside `set_group()` if it fails, and I understand it is like that on purpose because in examine() we use `set_group()` to test if the feature is supported -- and don't wish to print errors at that point.

So I'd recommend to put the error prints directly here into `halt_set_dcsr_ebreak()`.

There is one more occurrence of this same case at the bottom of this function.

---

set_group() only fails if dmi_write() or dmi_read() fails, and those do print errors (I'm fairly sure).

---

Actually, there are two failure modes for set_group():

- 1) ERROR_FAIL returned (due to DMI read/write error)
- 2) supported=false

I had the second failure mode in mind. So in case the removal from the halt group is not possible (for instance due to hardcoded values in the hardware - unlikely but possible), we would not be notified about that. That's what I recommend to cover by an error print - it may save some troubleshooting in the future.

---

Alright. This seems like a pretty extreme corner case.